proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,176 @@
+    Latency(4_000),
+    Latency(0_000),
+    // prometheus.LinearBuckets(10000, 10000, 5),
+    Latency(10_000),
+    Latency(20_000),
+    Latency(30_000),
+    Latency(40_000),
+    Latency(50_000),
+    // Prometheus implicitly creates a max bucket for everything that
+    // falls outside of the highest-valued bucket, but we need to
+    // create it explicitly.
+    Latency(u32::MAX),
+];
+
+/// A series of latency values and counts.
+#[derive(Debug)]
+pub struct Histogram([u32; NUM_BUCKETS]);
+
+/// A latency in milliseconds.
+#[derive(Debug, Default, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Hash)]
+pub struct Latency(u32);
+
+
+// ===== impl Histogram =====
+
+impl Histogram {
+
+    /// Observe a measurement
+    pub fn observe<I>(&mut self, measurement: I)
+    where
+        I: Into<Latency>,
+    {
+        let measurement = measurement.into();
+        let i = BUCKET_MAX_VALUES.iter()
+            .position(|max| &measurement <= max)
+            .expect("latency value greater than u32::MAX; this shouldn't be \
+                     possible.");
+        self.0[i] += 1;
+    }
+
+    /// Construct a new, empty `Histogram`.
+    ///
+    /// The buckets in this `Histogram` should mimic the Prometheus buckets
+    /// created by the Conduit controller's telemetry server, but with max
+    /// values one order of magnitude higher. This is because we're recording
+    /// latencies in tenths of a millisecond, but truncating these observations
+    /// to millisecond resolution.
+    pub fn new() -> Self {
+        Histogram([0; NUM_BUCKETS])
+    }
+
+}
+
+impl<I> ops::AddAssign<I> for Histogram
+where
+    I: Into<Latency>
+{
+    #[inline]
+    fn add_assign(&mut self, measurement: I) {
+        self.observe(measurement)
+    }
+
+}
+
+
+impl<'a> IntoIterator for &'a Histogram {
+    type Item = &'a u32;
+    type IntoIter = slice::Iter<'a, u32>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+
+}
+
+
+impl Default for Histogram {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// ===== impl Latency =====
+
+const SEC_TO_MS: u32 = 1_000;
+/// Conversion ratio from milliseconds to nanoseconds.
+pub const MS_TO_NS: u32 = 1_000_000;
+
+impl From<Duration> for Latency {
+    fn from(dur: Duration) -> Self {
+        let secs = dur.as_secs();
+        // checked conversion from u64 -> u32.
+        let secs =
+            if secs >= u64::from(u32::MAX) {
+                None
+            } else {
+                Some(secs as u32)
+            };
+        // represent the duration as ms.
+        let as_ms = {
+            let t = secs.and_then(|as_secs|
+                // convert the number of seconds to ms, or None on overflow.
+                as_secs.checked_mul(SEC_TO_MS)
+            );
+            let t = t.and_then(|as_ms| {
+                // convert the subsecond part of the duration (in ns) to ms.
+                let subsec_ms = dur.subsec_nanos() / MS_TO_NS;
```

Comment:
should this be TENTH_OF_MS_TO_NS (100_000)?

Reply:
I don't *think* so, I changed the `Latency` type to whole millisecond precision in afdf02bcc2f5a101959d3248d629e8c081664f96, since as we discussed, the truncation is no longer necessary?